### ISSUE TITLE
`needless_pass_by_ref_mut` should not trigger when values derived from the `&mut` are used in an unsafe block

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-clippy#13172]
### Summary

Using values derived from a `&mut` inside of an unsafe block can still trigger the `needless_pass_by_ref_mut` lint.

It seems like it was partially addressed in #11586, which should already be present in 1.80, but that only seems to cover direct use of the &mut in an unsafe block, not values derived from the &mut.

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code:

```rust
struct MyStruct {
    child: Option<NonNull<Self>>,
}

impl MyStruct {
    fn do_thing(&mut self) {
        if let Some(mut child) = self.child {
            unsafe { child.as_mut() }.do_thing()
        }
    }
}
```

I saw this happen:

```
this argument is a mutable reference, but not used mutably
for further information visit https://rust-lang.github.io/rust-clippy/master/
index.html#needless_pass_by_ref_mut
`-W clippy::needless-pass-by-ref-mut` implied by `-W clippy::nursery`
to override `-W clippy::nursery` add `#[allow(clippy::
needless_pass_by_ref_mut)]` (clippy needless_pass_by_ref_mut)
```

I expected to see this happen:

Nothing in this case. The compiler can't know if my unsafe usage of `child` is safe to call from a shared reference. Switching it to a shared reference could cause unsoundness.

### Version

```text
rustc 1.80.0 (051478957 2024-07-21)
binary: rustc
commit-hash: 051478957371ee0084a7c0913941d2a8c4757bb9
commit-date: 2024-07-21
host: x86_64-unknown-linux-gnu
release: 1.80.0
LLVM version: 18.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-bug

**Comment by alex-semenyuk** [comment]
Don't see this on 0.1.79  https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f2cdf7721eaa6a922e6447bd85b6389c

**Comment by ada4a** [comment]
fixed in #11624

@rustbot label S-fixed

**Comment by awused** [comment]
@ada4a Have you actually confirmed the specific example in the bug report is fixed?

I do not think that PR fixed this bug, since that PR only addresses direct use of the variable in an unsafe block, not values derived from the &mut. This bug actually mentions #11586, which was one of the bugs fixed in #11624, so I think this bug still exists.

Edit: Confirmed this bug still exists in `rustc 1.89.0 (29483883e 2025-08-04)` using the example code in the initial report.

**Comment by ada4a** [comment]
Oops, my bad -- I blindly took the Playground gist above, tried running it on latest stable, didn't see a warning, and therefore decided it must've been fixed -- but that was actually because the gist didn't contain a `#![warn(clippy::needless_pass_by_ref_mut)]` line... For the future, here's a gist with that line included: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6a80c6fdc26ebe3a96037432ce487deb

@rustbot label -S-fixed
